content/4_Gold/Combinatorics.mdx

```diff
@@ -0,0 +1,282 @@
+		 */
+		dp[i][0] = 1;
+		/*
+		 * i choose i is always 1 since there is exactly one way 
+		 * to choose i elements from a set of i elements
+		 * (choose every element in the set)
+		 */
+		if (k <= i) {
+			dp[i][i] = 1; 
+		}
+	}
+	for (int i = 0; i <= n; i++) {
+		for (int j = 1; j <= min(i, k); j++) {
+			if (i != j) {  // skips over the base cases
+				// uses the recurrence relation above
+				dp[i][j] = (dp[i - 1][j - 1] + dp[i - 1][j]) % p;
+			}
+		}
+	}
+	return dp[n][k];  // returns nCk modulo p
+}
+```
+</CPPSection>
+</LanguageSection>
+
+### Method 2: Factorial Definition (Modular Inverses) - $\mathcal{O}(n\log MOD)$
+
+First of all, we define $n!$ as $n \times (n - 1) \times (n - 2) \times \ldots 1$, which represents the number of permutations of a set of $n$ elements.
+
+Another way to calculate binomial coefficients is as follows:
+
+$$
+	\binom{n}{k} = \frac{n!}{k!(n-k)!}
+$$
+
+Since there are $n!$ permutations of $n$ elements, we go through all permutations and always include the first $k$ elements of the permutation in the subset. The order of the elements inside and outside the subset does not matter, so the result is divided by $k!$ and $(n − k)!$
+
+Since these binomial coefficients are large, problems typically require us to output the answer modulo a large prime $p$ such as $10^9 + 7$.
+
+Fortunately, we can use [modular inverses](https://usaco.guide/gold/modular?lang=cpp) to divide $n!$ by $k!$ and $(n - k)!$ modulo $p$. We can precompute all factorials in $\mathcal{O}(n)$ time and inverse factorials in $\mathcal{O}(n\log MOD)$ (although, not necessary, you can also precompute inverse factorials in $\mathcal{O}(n)$) time.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+		
+using namespace std;
+		
+const int MAXN = 1e6;
+
+long long fac[MAXN + 1], inv[MAXN + 1];
+		
+/** Computes x^y modulo p in O(log p) time. */
+
+long long exp(long long x, long long y, long long p) {
+	long long res = 1; x %= p;
+	while(y) {
+		if(y & 1) {
+			res *= x; res %= p; 
+		}
+		x *= x; x %= p;
+		y >>= 1;
+	}
+	return res;
+}
+
+/** Precomputes n! from 0 to MAXN. */
+		
+void factorial(long long p) {
+	fac[0] = 1;
+	for(int i = 1; i <= MAXN; i++) {
+		fac[i] = fac[i - 1] * i % p;
+	}
+}
+
+/** Precomputes all modular inverse factorials from 0 to MAXN in O(n log p) time */
+		
+void inverses(long long p) {
+	inv[0] = 1;
+	for(int i = 1; i <= MAXN; i++) {
+		inv[i] = exp(fac[i], p - 2, p);
+	}
+}
+		
+/** Computes nCr mod p */
+
+long long choose(long long n, long long r, long long p) {
+	return (fac[n] * inv[r] % p * inv[n - r] % p);
+}
+```
+</CPPSection>
+</LanguageSection>
+
+### Solution - Binomial Coefficients
+
+The first method for calculating binomial factorials is too slow for this problem since the constraints on $a$ and $b$ are $(1 \leq b \leq a \leq 10^6)$. However, we can use the second method to answer each of the $n$ queries in constant time by precomputing factorials in  and modular inverses.
```

Comment:
```suggestion
The first method for calculating binomial factorials is too slow for this problem since the constraints on $a$ and $b$ are $(1 \leq b \leq a \leq 10^6)$. However, we can use the second method to answer each of the $n$ queries in constant time by precomputing factorials and their modular inverses.
```